src/logger.ts

```diff
@@ -18,8 +18,19 @@ export interface OptionalLogger {
  */
 export type LogLevel = 'error' | 'info' | 'debug';
 
+export type ContextValue = unknown | {[key: string]: ContextValue};
+
+export type Context = {
+  [key: string]: ContextValue | undefined;
+};
+
 export interface LogSink {
   log(level: LogLevel, ...args: unknown[]): void;
+  /**
+   * If implemented, the LogSink will receive a Context object instead
+   * of the context being stringifying and prepended to args.
+   */
+  logWithContext?(level: LogLevel, context: Context, ...args: unknown[]): void;
```

Comment:
Personally I think we should keep things simple and break compat and only have a single log method that takes `context: Context|undefined, ...other args...`. We are the only real users of this library, we should make it work how we wnt.

Reply:
Sure, that would definitely make things simpler. I'll make a separate change and send it out. 

Tangentially, if we are changing the API, I would recommend renaming `addContext()` to `withContext()`, given that it's an immutable object API that returns a new one. (I was confused when I first looked at the code because "add" usually changes the object).

That would result in more search-and-replace, but I thought I'd put it out there while we're changing the API.

Reply:
Your name change makes sense, go for it.